installer image, integrity protection

pohly opened the pull request:
This PR is now ready for a final review. Actually building it is currently broken, most likely because the oe-selftest support in the CI system needs to be fixed first.

@ipuustin can you help review the security side of this?

@gvancuts @dbkinder can you help reviewing the new documentation (doc/security.rst and the new section in doc/howtos/image-install.rst)

okartau commented:
can submodule url be changed to https, so that builder can fetch it without using git:// as it does with other submodules

pohly commented:
test this please

pohly commented:
test this please

okartau commented:
CI tester does not find image because image file name is formed differently.
tester expects:
refkit-image-gateway-intel-corei7-64-2017-03-30_08-15-23-build-539.wic.xz
image file built:
refkit-image-gateway-intel-corei7-64-20170330155801.wic.xz

pohly commented:
On Thu, 2017-03-30 at 09:42 -0700, okartau wrote:
> CI tester does not find image because image file name is formed
> differently.
> tester expects:
> refkit-image-gateway-intel-corei7-64-2017-03-30_08-15-23-build-539.wic.xz
> image file built:
> refkit-image-gateway-intel-corei7-64-20170330155801.wic.xz

Does the post-build test include refkit-ci.inc in auto.conf?

There is a suspicious 10 minute delay when testing starts, for example:

00:42:32.434 2017-03-30 15:57:53,559 - selftest - INFO - Adding: "include bblayers.inc" in bblayers.conf
00:42:32.434 2017-03-30 15:57:53 - test_install_fixed_password (oeqa.selftest.image-installer.ImageInstaller)
00:52:42.780 Test image installation under qemu without virtual TPM, using a fixed password ... Loading cache...done.
00:52:42.780 rootfs file: /srv/jenkins/workspace/builder-slot-0/build/tmp-glibc/deploy/images/intel-corei7-64/refkit-installer-image-intel-corei7-64.wic

That test setup should be faster.

My theory is that actual build and post-build testing are not set up
alike enough, causing the file names to change, and therefore all images
get generated anew. The default in OE-core is to remove old images
(actually, I think only that is supported anymore), so that would
explain why the expected images are gone.

pohly commented:
On Thu, 2017-03-30 at 21:13 +0200, Patrick Ohly wrote:
> On Thu, 2017-03-30 at 09:42 -0700, okartau wrote:
> > CI tester does not find image because image file name is formed
> > differently.
> > tester expects:
> > refkit-image-gateway-intel-corei7-64-2017-03-30_08-15-23-build-539.wic.xz
> > image file built:
> > refkit-image-gateway-intel-corei7-64-20170330155801.wic.xz
> 
> Does the post-build test include refkit-ci.inc in auto.conf?
> 
> There is a suspicious 10 minute delay when testing starts, for example:
> 
> 00:42:32.434 2017-03-30 15:57:53,559 - selftest - INFO - Adding: "include bblayers.inc" in bblayers.conf
> 00:42:32.434 2017-03-30 15:57:53 - test_install_fixed_password (oeqa.selftest.image-installer.ImageInstaller)
> 00:52:42.780 Test image installation under qemu without virtual TPM, using a fixed password ... Loading cache...done.
> 00:52:42.780 rootfs file: /srv/jenkins/workspace/builder-slot-0/build/tmp-glibc/deploy/images/intel-corei7-64/refkit-installer-image-intel-corei7-64.wic
> 
> That test setup should be faster.
> 
> My theory is that actual build and post-build testing are not set up
> alike enough, causing the file names to change, and therefore all images
> get generated anew. The default in OE-core is to remove old images
> (actually, I think only that is supported anymore), so that would
> explain why the expected images are gone.

I think I found it:

docker/build-project.sh:export BUILD_ID=${CI_BUILD_ID}

I bet that part is missing for oe-selftest invocations, and thus
CI_BUILD_ID gets ignored, causing exactly the behavior I described
above.

pohly commented:
test this please

pohly commented:
retest this please
test this please

okartau commented:
test this please

okartau commented:
it's good to have BUILD_ID differently in test scripts than build scripts, I explained it in bug about BUILD_ID, so that part should be removed from this patch set. 
A post-build tester should not become builder.

pohly commented:
test this please

pohly commented:
test this please

pohly commented:
test this please

pohly commented:
Rebased and removed as many local workarounds as possible in the foreseeable future. The remaining ones (like enabling the swtpm spawning through qemu) will take longer to resolve upstream.

pohly commented:
test this please

pohly commented on doc/howtos/image-install.rst:
```diff
@@ -86,3 +86,52 @@ includes a checksum verification, you can also use the traditional :command:`dd`
 Unplug the removable media from your development system and you're ready to plug
 it into your target system.
+
+Installing to internal Media
```
I did that here to be consistent with capitalization in headers elsewhere ("major" words in uppercase, adjectives lowercase - not sure whether that's exactly the rule).

I don't have a strong preference regarding media vs. medium. I used singular when talking about one particular instance and plural when there could be more than one.

pohly commented on doc/howtos/image-install.rst:
```diff
@@ -86,3 +86,52 @@ includes a checksum verification, you can also use the traditional :command:`dd`
+============================
+
+With the approach above, the OS always boots from the removable
+media. To install the OS to the internal media, one can build the
```
Makes sense.

pohly commented on doc/howtos/image-install.rst:
```diff
@@ -86,3 +86,52 @@ includes a checksum verification, you can also use the traditional :command:`dd`
+be created which automatically installs the OS.
+
+There are also example scripts under ``doc/howtos/image-installer``
+that show how to do installation with a TPM in a qemu virtual
```
Yes.

pohly commented on doc/howtos/image-install.rst:
```diff
@@ -86,3 +86,52 @@ includes a checksum verification, you can also use the traditional :command:`dd`
+There are also example scripts under ``doc/howtos/image-installer``
+that show how to do installation with a TPM in a qemu virtual
+machine. On a build host, create a build environment for building
+without Docker as explained in the README.rst, then enter::
```
Okay. I simply wasn't sure how to link to other documentation.

pohly commented on doc/security.rst:
```diff
@@ -0,0 +1,184 @@
+Real products must use their own key. Those can be created for
+example with::
+
+  openssl genrsa -aes128 -passout pass:<your key password> -out private.pem 4096
```
No particular reason. I just hadn't noticed that particular aspect ;-} I'll try genpkey.

pohly commented on doc/security.rst:
Okay.

pohly commented on doc/security.rst:
Okay.

ipuustin commented on meta-refkit/classes/image-installer.bbclass:
```diff
@@ -0,0 +1,370 @@
+# image.
+#
+# Basically that turns the image using image-installer.bbclass into an
+# "installer image" which can be booted from a removal media to
```
removal -> removable

ipuustin replied:
Btw, commit msg says "Ostro OS OX"?

ipuustin commented on meta-refkit/classes/image-installer.bbclass:
```diff
@@ -0,0 +1,370 @@
+
+# Cleanup handling, common code for selecting among different input and output
+# options interactively, etc.
+INSTALLER_UTILS[shellcheck] = "sh"
```
Did you run this through shellcheck? It reports ~8 warnings and style issues, nothing major though.

pohly replied:
I did at some point, but not recently. The INSTALLER_LOGGING[shellcheck] annotation is for some non-upstream OE-core extension of the verify-bashism script, to be used with "verify-bashisms initramfs-framework-refkit-luks refkit-installer-image", which then runs shellcheck in addiditon to checking for bashisms.

ipuustin commented on meta-refkit/classes/image-installer.bbclass:
```diff
@@ -0,0 +1,370 @@
+            # rooted in.
+            local disk=
+            # shellcheck disable=SC2034
+            exclude=$( lsblk --output NAME,KNAME,TYPE --ascii --noheadings | while read name kname type; do
```
Is `lsblk` assumed to be provided by the host?

pohly replied:
That's for the target, covered by:

    INSTALLER_RDEPENDS ??= " \
        util-linux \
    "

ipuustin commented on meta-refkit/recipes-image/images/refkit-installer-image.bb:
```diff
@@ -0,0 +1,188 @@
+INSTALLER_INSTALL ?= "${REFKIT_INSTALLER_UEFI_COMBO}"
+
+INSTALLER_RDEPENDS_append = " \
+    dosfstools \
```
GPLv3

ipuustin commented on meta-refkit/recipes-image/images/refkit-installer-image.bb:
```diff
@@ -48,10 +52,20 @@ REFKIT_INSTALLER_UEFI_COMBO () {
         done
         if [ ! "$partition" ]; then
-            fatal "could not identify partition #$gdisk_pnum in $output"
+            fatal "could not identify parition #$gdisk_pnum in $output"
```
What's this?

ipuustin commented on meta-refkit/recipes-image/images/refkit-installer-image.bb:
```diff
@@ -23,10 +23,14 @@ REFKIT_INSTALLER_UEFI_COMBO () {
         output_mountpoint=
+        output_luks=
+        LUKS_NAME=rootfs
+        LUKS_PASSWORD=refkit
```
There should probably be a big warning message so that no-one will use this password accidentally.

pohly replied:
The password feature is always enabled, but never used unless explicitly enabled by setting FIXED_PASSWORD=yes when calling image-installer.

Are you suggesting to not put the code into the image-installer script by default and once enabled, print a warning? That seems counter-intuitive - after all, the feature would have been enabled intentionally in such a setup.

ipuustin replied:
Ok, it wasn't clear to me from the commit that the password is not used by default. It just worries me a lot whenever I see a hardcoded password in the sources. :-) But are the `FIXED_PASSWORD` and `LUKS_PASSWORD` settings overlapping a bit, if one is used only if another is enabled? I mean, could we just say `FIXED_PASSWORD="refkit" image-installer`?

pohly replied:
LUKS_PASSWORD is basically an internal constant. Its value has to match the string in initramfs-framework-refkit-luks.

FIXED_PASSWORD=yes is what enables the use of that fixed password.

We could make it so that there is no default password in initramfs-framework-refkit-luks unless the developer configures one in his build, and the image-installer script could be changed so that this fixed password has to passed via the FIXED_PASSWORD env variable. Would that be better? It's harder to use, but perhaps that is how we want it to be ;-}

pohly replied:
I've made it harder to accidentally use the fixed password, see the new "installer image: fixed password support disabled by default" commit.

ipuustin commented on meta-refkit/recipes-image/images/files/initramfs-framework-refkit-luks:
```diff
@@ -0,0 +1,52 @@
+            cryptsetup isLuks "$bootparam_root" 2>/dev/null
+            case $? in
+              0) # LUKS volumne
+                if echo "$LUKS_PASSWORD" | cryptsetup open --type luks "$bootparam_root" "$LUKS_NAME" --key-file -; then
```
What's the trailing `-`? Is it for reading the password?

pohly replied:
Yes, it makes cryptsetup read the password from stdin.

mythi commented on meta-refkit/recipes-image/images/refkit-installer-image.bb:
```diff
@@ -0,0 +1,188 @@
+            fatal "sgdisk $output has failed - damaged disk?"
+        fi
+
+        # Read partition description from rootfs.
```
Can we get rid of this emmc-partitions-data dependency? Currently, it's essentially ```DSK_IMAGE_LAYOUT``` with ```INT_STORAGE_ROOTFS_PARTUUID_VALUE``` replaced as the rootfs PARTUUID.

However, now we need to keep ```DSK_IMAGE_LAYOUT``` up-to-date with .wks.in just to get emmc-partitions-data correct.  Since the goal with our installer is just "take all but rootfs as is but make rootfs take all the space on the output device and make its ```PARTUUID==INT_STORAGE_ROOTFS_PARTUUID_VALUE```, could we do this runtime and just have ```INT_STORAGE_ROOTFS_PARTUUID_VALUE``` substituted when processing the script?

pohly replied:
Is INT_STORAGE_ROOTFS_PARTUUID_VALUE guaranteed to be the same for all images that the script will be asked to install? If yes, then it would be possible to embed the value directly in the script.

But probably that is an oversimplification, because technically INT_STORAGE_ROOTFS_PARTUUID_VALUE is a per-image variable and could be set differently.

mythi replied:
True, it cannot be guaranteed and we should not make that limitation.  I had not checked this carefully enough and still had the mindset (don't know where that came to me) that the installer image had just the rootfs'es of the other images in it (and not the full images).

Anyway, the likely mismatch between ```DSK_IMAGE_FORMAT``` and what's in .wks.in is a problem. How about if do_uefiapp copies the "internal storage" .efi in ```EFI_<PARTUUID>``` and the rootfs PARTUUID is parsed from it.

With, e.g., ```sgdisk -p <wic>``` we can get the same partition data and the partitions carry the filesystem info.

pohly replied:
@mythi I'm not exactly following what you are suggesting. Can we keep this as it is now and clean it up once the code gets reorganized (do_uefiapp -> meta-intel, etc.)?

mythi replied:
We can clean it up later.

What I'm suggesting is:
- the image partition layout is read from the image and not from emmc_partitions_data
- the install target rootfs PARTUUID is read from the EFI directory name  (currently it's EFI_Internal_storage but could become ```EFI_<PARTUUID>```). Needs a change in do_uefiapp.

ipuustin commented on doc/howtos/image-install.rst:
```diff
@@ -86,3 +86,52 @@ includes a checksum verification, you can also use the traditional :command:`dd`
+
+By default, the installer image contains different images that can be
+installed, as configured by the ``INSTALLER_SOURCE_IMAGES`` variable.
+It uses dm-verity to ensure that the installer image did not get
```
dm-verity -> dm-verify

ipuustin replied:
No, sorry, it was dm-verity. Thought this was a typo. :-)

ipuustin commented on doc/howtos/image-installer/init-tpm:
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh -ex
```
Might it be a good idea to have these scripts in `intel-iot-refkit/scripts/image-installer` instead of `intel-iot-refkit/doc/howtos/image-installer`?

pohly replied:
I don't have a strong opinion about the location of these scripts. I can move them.

ipuustin commented on doc/security.rst:
```diff
@@ -0,0 +1,184 @@
+A real product also needs to ensure that the device boots with an
+initramfs which really activates dm-verity.
+:file:`meta-refkit/recipes-image/images/initramfs-framework-refkit-dm-verity.bb`
+in ``refkit-initramfs`` does that when ``dmverity`` is part of the
```
This is a bit confusing part. The initramfs module doesn't ensure that we are booting from an initramfs which activates dm-verity. It checks the dmverity signature instead.

pohly replied:
The documentation refers to this line:

          if [ -e "$bootparam_root" ] && [ -e "$bootparam_dmverity" ]; then

It's the dmverity boot parameter which must be present to trigger the code, which must be set when building the image. Would it be better to say:

> A real product also needs to ensure that its image really activates dm-verity. This is done by
> booting with the ``dmverity`` boot parameter, which tells :file:`meta-refkit/recipes-image/images/initramfs-framework-refkit-dm-verity.bb`
> in ``refkit-initramfs`` that dm-verity is required. :file:`refkit-image.bbclass` automatically adds that boot parameter when
> building an image with dm-verity enabled. Secure Boot is required to prevent attackers from replacing the initramfs 
> and/or changing boot parameters.

ipuustin replied:
I think this is clearer.

ipuustin commented on doc/security.rst:
This sentence is difficult for me to parse.

ipuustin commented on doc/security.rst:
We should maybe copy that documentation?

ipuustin commented:
The meta-openembedded update in this PR is intentional?

ipuustin commented on meta-refkit/classes/image-installer.bbclass:
```diff
@@ -0,0 +1,370 @@
+# already by the bitbake parser. The downside is that we are limited
+# to shell code that the bitbake parser understands (no bash
+# extensions, for example) and error reports about the shell syntax
+# are a bit hard to read.
```
I think the question is whether embedding the code inside recipe files makes it less maintainable in the long run. For example, how do you write tests for it?

pohly replied:
How would one write tests if it was a separate file? I don't think that would be different. In both cases one needs to boot into the installer image and then exercise the script in various ways. This can be done via oe-selftest.

Whether embedding vs. separate file is easier from a maintenance perspective is debatable. I found it quite useful.

ipuustin replied:
I didn't mean functional testing ("Does the installer work?"), but rather unit- or module-level tests ("Does pick_option() fail gracefully with these inputs?"). There are frameworks for testing shell scripts, like https://github.com/kward/shunit2. Since we don't have any tests for this it doesn't matter much at the moment, but the people customizing their own installer scripts might want to do it. This is not a showstopper though...

ipuustin commented on meta-refkit/classes/image-installer.bbclass:
```diff
@@ -0,0 +1,370 @@
+# See below for the commands used by each step.
+#
+# The default installer code is free of bashisms and passed
+# checks with checkbashisms.pl and checkshell. The "local"
```
checkshell -> shellcheck ?

ipuustin commented on meta-refkit/classes/image-installer.bbclass:
```diff
@@ -0,0 +1,370 @@
+# The default installer code is free of bashisms and passed
+# checks with checkbashisms.pl and checkshell. The "local"
+# keyword (although supported by bash and dash) is avoided
+# in favor of wrapping function bodies with local variables
```
This says `local` is avoided, but I still see several variables with `local` keyword...

pohly replied:
I think I added that back later... probably should remove "local" again.

pohly commented on doc/security.rst:
Perhaps. @mythi, are you working on enhancing support for Secure Boot in refkit?

pohly commented:
On Wed, 2017-04-05 at 00:38 -0700, Ismo Puustinen wrote:
> The meta-openembedded update in this PR is intentional?

Yes. It contains some of my fixes that are needed for this PR.

ipuustin commented on meta-refkit/classes/image-installer.bbclass:
```diff
@@ -0,0 +1,370 @@
+
+    # Checks whether a variable is "yes" or "no", then returns
+    # with 0 when "yes".
+    istrue () {
```
I guess `if [ "x$1" = "xyes" ];` won't cut it?

pohly replied:
That's what I had initially, but then typos like x=yess aren't caught.

mythi commented on doc/security.rst:
Yes I am. I saw this section and will be updating that part with what I have.

ipuustin commented on meta-refkit/classes/refkit-hash-dm-verity-key.bbclass:
```diff
@@ -0,0 +1,28 @@
+        bb.fatal('REFKIT_DMVERITY_PRIVATE_KEY=%s is not a file.' % privkey)
+    with open(privkey, 'rb') as f:
+        data = f.read()
+    hash = hashlib.sha256(data).hexdigest()
```
Will fail or calculate wrong hash if file can't be opened? Indent to be part of the `with open...` block and/or check `data`.

pohly replied:
It'll bail out with an exception if open() or read() fail, so data and hash are always valid here when execution continues. Indention doesn't matter. I prefer it this way because the hash operation doesn't depend on the file still being open.

ipuustin replied:
Ok. The code comment just talks handling the other error cases, but not these.

ipuustin commented on meta-refkit/recipes-image/images/initramfs-framework-refkit-dm-verity.bb:
```diff
@@ -0,0 +1,112 @@
+
+    dmverity_run () {
+        C=0
+        delay=${bootparam_rootdelay:-1}
```
Should you check for `bootparam_rootwait` too?

pohly replied:
initramfs-framework doesn't use that parameter elsewhere either. It might make sense to allow waiting indefinitely (because the fallback is to give up), but that should be introduced across the board, not just here.

ipuustin commented on meta-refkit/recipes-image/images/refkit-installer-image.bb:
```diff
@@ -0,0 +1,270 @@
+                fi
+            else
+                if ! execute sgdisk -n "$gdisk_pnum:+0:+${size}M" -c "$gdisk_pnum:$lname" \
+                       -t "$gdisk_pnum:$type_id" -u "${gdisk_pnum}:${uuid}" -- "$output"; then
```
I'm bit confused about using both `$gdisk_pnum` and `${gdisk_pnum}` on this line.

pohly replied:
That was copied verbatim from https://github.com/otcshare/ostro-os-xt/blob/master/meta-ostro-xt/recipes-xt/initrdscripts/initramfs-framework-installer/installer#L133

I think it is safe (and better) to replace ${gdisk_pnum} with $gdisk_pnum and ${uuid} with $uuid.

ipuustin replied:
I agree. My first thought was that it was some BB substitution.

pohly commented on doc/security.rst:
Is this better:

> The key remains in kernel memory and it is the kernel's responsibility to
> not leak it during a runtime attack. In addition, the kernel must prevent
> an attacker from modifying files at runtime, because such changes cannot
> be detected during the next boot.

ipuustin commented on doc/security.rst:
That's much better.

pohly commented:
Another change - "use_encryption" was only documented in the commit message. It's now also in security.rst:

    diff --git a/doc/security.rst b/doc/security.rst
    index 32f2510..8d84772 100644
    --- a/doc/security.rst
    +++ b/doc/security.rst
    @@ -116,6 +116,7 @@ Currently, `LUKS`_ is used for the rootfs partition. Using LUKS has
     several advantages:
     
     * ``refkit-initramfs`` can detect whether a partition is encrypted
    +  and uses the partition accordingly
     
     * LUKS tools have sane default parameters for encryption
     
    @@ -132,6 +133,15 @@ out-of-the-box.
     
     .. _LUKS: https://gitlab.com/cryptsetup/cryptsetup/blob/master/README.md
     
    +By default, ``refkit-initramfs`` does not enforce the use of
    +encryption. That is because the images are meant to be usable both
    +after flashing to removable media (no encryption) and after
    +installation (optionally with encryption). If encryption is meant to
    +provide integrity protection, booting without encryption should be
    +disabled by adding the ``use_encryption`` boot parameter to the
    +``APPEND`` variable of an real product image. Otherwise an attacker
    +could replace the entire rootfs and then boot into that.
    +
     To unlock the LUKS partition, a secret key is
     needed. ``image-installer`` and the corresponding
     :file:`meta-refkit/recipes-image/images/initramfs-framework-refkit-luks.bb`
    diff --git a/openembedded-core b/openembedded-core
    --- a/openembedded-core
    +++ b/openembedded-core
    @@ -1 +1 @@
    -Subproject commit d68a86d87aa017dd0fecb0f626d22711efefcd3f
    +Subproject commit d68a86d87aa017dd0fecb0f626d22711efefcd3f-dirty

gvancuts commented:
Thanks @pohly 👍

pohly commented:
test this please

okartau commented:
test this please

pohly commented:
Last test run failed due to an unrelated problem:

    AssertionError: 1 != 0 : Errors in boot log:
     Mar 30 10:29:24 intel-corei7-64 kernel: [drm:intel_dp_aux_ch] *ERROR* dp_aux_ch not done status 0xa1450064,

rojkov commented:
would be nice if the PR included meta-openembedded update up to b0699e4. Then I wouldn't need to update the layer in my upcoming PR which drops python2 from all images.

pohly commented:
@mythi said on IRC that he wants to wait for a refkit snapshot to be tagged before merging, so I'm continuing to investigate the image rebuilding during post-selftest here.

@rojkov I've update meta-openembedded to latest master.

pohly commented:
test this please

pohly commented:
test this please

pohly commented:
Third times the charm... I hope.

test this please

pohly commented:
test this please

okartau commented:
test this please

pohly commented:
test this please

pohly commented:
test this please

pohly commented:
Finally managed to get a bitbake-diffsig between do_rootfs from the main build and post testing (note for future reference: when building from sstate, there is no sigdata file anywhere for that build; one has to use "bitbake -S none").

It is the buildhistory which causes the image rebuilding. From https://ostroproject.org/jenkins/job/intel-iot-refkit_pull-requests/631/consoleFull:

00:33:54.247 + bitbake-diffsigs tmp-glibc/stamps/intel_corei7_64-refkit-linux/refkit-initramfs/1.0-r0.do_rootfs.sigdata.ae662fb357580faa04a25a310e02e77a tmp-glibc/stamps/intel_corei7_64-refkit-linux/refkit-initramfs/1.0-r0.do_rootfs.sigdata.bc77000cdcfa37762afe751ef8cd55b9
00:33:54.247 Task dependencies changed from:
...
00:33:54.247 changed items: {'buildhistory_get_imageinfo'}
...
00:33:54.247 Dependency on variable buildhistory_get_build_id was added
00:33:54.247 Dependency on variable buildhistory_list_installed was added
00:33:54.247 Dependency on variable buildhistory_get_imagevars was added
...
00:33:54.247 Variable IMAGE_POSTPROCESS_COMMAND value changed from 'write_deploy_manifest; ' to 'write_deploy_manifest;   buildhistory_get_imageinfo ; '
00:33:54.247 Variable ROOTFS_POSTUNINSTALL_COMMAND value changed from 'write_image_manifest ;  stateless_mangle_rootfs;' to 'write_image_manifest ;    buildhistory_list_installed_image ;                                buildhistory_get_image_installed ; stateless_mangle_rootfs;'

pohly commented:
test this please

Removed the debug commit as the root cause is now understood, updated meta-security to current master.

Can be merged as-is if testing with newer meta-security succeeds.

pohly commented:
test this please

pohly commented:
test this please

pohly commented:
Build and oe-selftest worked, but aft testing didn't ("unstable" result?)

test this please

mythi commented:
On Wed, Apr 12, 2017 at 11:49 PM, Patrick Ohly <notifications@github.com>
wrote:

> Build and oe-selftest worked, but aft testing didn't ("unstable" result?)
>
>
> unstable means a glitch in DAFT itself (or something with the tester
boards)

skuusela commented:
Should be fixed now, test this please

pohly commented:
Several new commits in this revision of the PR:

* swtpm: bump SRCREV (temporary hack) a232869  - can be removed after next meta-security update
* qemu: replace CUSE with socket backend e270c71 - permanent, patches still under review and need updating, so not suitable for OE-core yet
* refkit-image.bbclass: use virtio under runqemu   47a6af3
* CI Jenkinsfile: pass host /dev/kvm into docker session  24a7c17
* refkit-ci.inc: conditionally enable the use of KVM   38bc505

This is working. The latest test build uses KVM and virtio.

@mythi: ready for merging, IMHO.
